src/module.js

```diff
@@ -16,12 +16,18 @@ export default function Module(runtime) {
 Object.defineProperties(Module.prototype, {
   _copy: {value: module_copy, writable: true, configurable: true},
   _resolve: {value: module_resolve, writable: true, configurable: true},
+  redefine: {value: module_redefine, writable: true, configurable: true},
   define: {value: module_define, writable: true, configurable: true},
   derive: {value: module_derive, writable: true, configurable: true},
   import: {value: module_import, writable: true, configurable: true},
   variable: {value: module_variable, writable: true, configurable: true}
 });
 
+function module_redefine(name) {
+  var v = this._scope.get(name) || this;
+  return v.define.apply(v, arguments);
+}
+
```

Comment:
It feels like a nicer API (from the JavaScript consumption side, not from our notebook editing side) — would be to call this function `define`. It takes a name, and either defines a variable of that name if one does not already exist, or redefines the existing variable by the same name.

We could then call the current `define`, `declare` or `assign` or `defineStrict`, and use it internally to make sure that you don't have cells with duplicate names in a notebook context...

Reply:
An interesting topic!

Changing the behavior of *module*.define would not be backwards-compatible, so we’d need a major version bump of the runtime (to 3.0.0). I’m willing to make breaking changes if necessary, but first allow me to argue that “redefine” is the right name here.

This method is intended only to allow changing the definition of an existing variable, not defining a new variable. I’ve updated the implementation so that it conforms more precisely to my intent: it will now throw a RuntimeError if you attempt to redefine an undefined variable. The goal here is not to enable arbitrary programming (that’s what the existing *module*.variable *etc.* methods are for); it’s to provide a convenient way to override an existing program.

A challenge here is that there are two concepts of symbolic identity. From the runtime’s perspective, it’s the variable *instance* that defines identity. In an Observable notebook, each cell corresponds to one or more variable. Thus, it’s necessary to allow multiple variables with the same name, which causes a duplicate definition error at runtime. But from the author’s perspective, you want to use variable *names* as identity. Meaning, you want to redefine the existing *height* variable rather than create a duplicate variable causing both to error.

I also considered exposing a *module*.resolve method that unlike the existing *module*._resolve throws an error if a variable with the specified *name* does not exist. But that would look cause redefinition to be more verbose since the variable’s *name* is not part of its identity:

```js
main.resolve("height").define("height", 70);
```

We could rename *module*.redefine to *module*.assign, analogous to an assignment in JavaScript that throws a ReferenceError in strict mode.

```js
main.assign("height", 70);
```

But the concept of an imperative assignment feels a little out of place in a dataflow programming model (even though, by nature of being embedded within an imperative language, that’s what this is). And that raises the question of whether *module*.assign would take a literal value or a reactive definition (with named inputs and an arbitrary function to be called whenever those inputs change), as the current implementation of *module*.redefine does. I’m inclined to support reactivity both for consistency and flexibility, so I think “redefine” really is the right name.

Reply:
Very well argued! 

But I think that ultimately, we are, in fact — if not trying to enable arbitrary programming at the main module level here — at least trying to enable clean, convenient, arbitrary integration between an Observable Notebook and your arbitrary JavaScript application.

To that end, I think we want to provide the most fundamental, useful, well-named primitives we can at the notebook level. Here's one more idea:

Assuming that `main` is the top-level "notebook" or "main module" value returned from loading or instantiating a new Observable runtime...

To determine whether or not a notebook has a cell with a given name:
```js
main.has(name)
```

To remove a cell of a given name from a notebook:
```js
main.delete(name)
```

To "redefine" (as above) an existing cell, or create a new cell with a definition, if no cell of that name yet exists (as a literal value, or as a reactive definition):
```js
main.set(name, definition)
```

To either get the most recently computed value of a given cell ... or perhaps to get a reference to the cell's `variable` object itself ... I'm not sure which is more interesting:
```js
main.get(name)
```

To get an array or Iterator of all of the names defined in the notebook:
```js
main.keys()
```

To disconnect and dispose of all of the notebook's cells at once:
```js
main.clear()
```

You see what I'm going for — it's notebook API as ES6 Map analogy.

Reply:
I’m not arguing against arbitrary programming; I’m saying: we already have an API for reactive programming (*module*.variable, *variable*.define, *etc.*); that API is oriented around symbolic *variables* whose identity is determined by instance rather than by name; and I don’t think we want a second, competing API that is oriented around names.

I’m offering a limited out here to redefine an existing variable because it’s a convenient way of passing values into an Observable program at runtime (similar to import-with). If an author wants more than that, she should either using the existing runtime API for meta-programming, or edit her Observable program directly.

Reply:
To illustrate a little further how the runtime is oriented around variables, consider what happens when you have a module that has more than one variable with the same name: each conflicting variable rejects with a RuntimeError. And *module*.redefine doesn’t change that, and it’s not appropriate to adopt the ES map analogy when names are neither required nor unique.

(Also, I fixed the behavior so that *module*.redefine now throws an error in this case rather than silently having no effect.)